File: engine/backend/src/entity-repository.ts
import type { EntityId, ComponentId } from "@engine/core";

export class EntityRepository {
  private readonly entities = new Set<EntityId>();
  private nextId = 0;

  /** entityId → set of direct componentIds */
  private readonly components = new Map<EntityId, Set<ComponentId>>();

  create(): EntityId {
    const id = `${Date.now()}-${this.nextId++}` as EntityId;
    this.entities.add(id);
    this.components.set(id, new Set());
    return id;
  }

  delete(id: EntityId): boolean {
    this.components.delete(id);
    return this.entities.delete(id);
  }

  has(id: EntityId): boolean {
    return this.entities.has(id);
  }

  getAll(): ReadonlySet<EntityId> {
    return this.entities;
  }

  addComponent(entityId: EntityId, componentId: ComponentId): void {
    const entityComponents = this.components.get(entityId);
    if (!entityComponents) {
      throw new Error(`Entity ${entityId as string} does not exist`);
    }
    if (entityComponents.has(componentId)) {
      throw new Error(
        `Component ${componentId as string} already exists on entity ${entityId as string}`,
      );
    }
    entityComponents.add(componentId);
  }

  removeComponent(entityId: EntityId, componentId: ComponentId): boolean {
    const entityComponents = this.components.get(entityId);
    if (!entityComponents) return false;
    return entityComponents.delete(componentId);
  }

  hasComponent(entityId: EntityId, componentId: ComponentId): boolean {
    return this.components.get(entityId)?.has(componentId) ?? false;
  }

  getComponentIds(entityId: EntityId): ComponentId[] {
    const set = this.components.get(entityId);
    return set ? [...set] : [];
  }
}

File: engine/backend/EntityRepository.cs
using Engine.Core;

namespace Engine.Backend;

public class EntityRepository
{
    private readonly HashSet<EntityId> _entities = new();
    private readonly Dictionary<EntityId, HashSet<string>> _components = new();
    private int _nextId;

    public EntityId Create()
    {
        var id = new EntityId($"{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}-{_nextId++}");
        _entities.Add(id);
        _components[id] = new HashSet<string>();
        return id;
    }

    public bool Delete(EntityId id)
    {
        _components.Remove(id);
        return _entities.Remove(id);
    }

    public bool Has(EntityId id) => _entities.Contains(id);

    public IReadOnlySet<EntityId> GetAll() => _entities;

    public void AddComponent(EntityId entityId, string componentId)
    {
        if (!_components.TryGetValue(entityId, out var set))
            throw new InvalidOperationException($"Entity {entityId} does not exist");
        if (!set.Add(componentId))
            throw new InvalidOperationException(
                $"Component {componentId} already exists on entity {entityId}");
    }

    public bool RemoveComponent(EntityId entityId, string componentId)
    {
        if (!_components.TryGetValue(entityId, out var set))
            return false;
        return set.Remove(componentId);
    }

    public bool HasComponent(EntityId entityId, string componentId)
    {
        return _components.TryGetValue(entityId, out var set) && set.Contains(componentId);
    }

    public IReadOnlyList<string> GetComponentIds(EntityId entityId)
    {
        if (!_components.TryGetValue(entityId, out var set))
            return Array.Empty<string>();
        return set.ToList();
    }
}
